core/trino-main/src/main/java/io/trino/cost/TopNStatsRule.java

```diff
@@ -47,8 +48,16 @@ protected Optional<PlanNodeStatsEstimate> doCalculate(TopNNode node, StatsProvid
         PlanNodeStatsEstimate sourceStats = statsProvider.getStats(node.getSource());
         double rowCount = sourceStats.getOutputRowCount();
 
-        if (node.getStep() != TopNNode.Step.SINGLE) {
-            return Optional.empty();
+        /* CreatePartialTopN rule runs after ReorderJoins but before DetermineJoinDistributionType and DetermineSemiJoinDistributionType.
+         * Therefore, it is useful to provide estimates for partial and final topN nodes.
+         * If, for example, the partial topN is part of a source stage, then it's overall output will depend on the number of splits that get created at runtime,
+         * and that's not known in advance. We populate a rough estimate for partial topN by assuming an input of 1 million rows per driver.
+         */
+        if (node.getStep() == TopNNode.Step.PARTIAL) {
+            double estimatedOutputRowCount = Math.max(rowCount / ESTIMATED_PARTIAL_TOPN_INPUT_PER_DRIVER, 1) * node.getCount();
```

Comment:
Do you need to explicitly protect against `rowCount` being unknown?

Reply:
not required, with current code we get no estimate for partial topN and node#getCount estimate for final topN as expected.